File: practice_mvvm/Paging/app/src/main/java/com/miraclehwan/paging/view/MainActivity.java
package com.miraclehwan.paging.view;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.arch.paging.PagedList;
import android.databinding.DataBindingUtil;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;

import com.miraclehwan.paging.R;
import com.miraclehwan.paging.adapter.ItemAdapter;
import com.miraclehwan.paging.databinding.ActivityMainBinding;
import com.miraclehwan.paging.model.Item;
import com.miraclehwan.paging.viewmodel.ItemViewModel;

public class MainActivity extends AppCompatActivity {

    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        binding.rvAnswers.setLayoutManager(new LinearLayoutManager(this));
        final ItemAdapter itemAdapter = new ItemAdapter(this);

        ItemViewModel itemViewModel = ViewModelProviders.of(this).get(ItemViewModel.class);

        itemViewModel.itemPagedList.observe(this, new Observer<PagedList<Item>>() {
            @Override
            public void onChanged(@Nullable PagedList<Item> items) {
                itemAdapter.submitList(items);
            }
        });

        binding.rvAnswers.setAdapter(itemAdapter);
    }
}

File: Paging3/app/src/main/java/com/miraclehwan/paging/test/MainActivity.kt
package com.miraclehwan.paging.test

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.miraclehwan.paging.R
import com.miraclehwan.paging.databinding.ActivityMainBinding
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    val viewModel by lazy { ViewModelProvider(this)[MainViewModel::class.java] }

    val mainListAdapter = MainAdapter()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = DataBindingUtil.setContentView<ActivityMainBinding>(this,
            R.layout.activity_main
        )

//        viewModel.listData.subscribe {
//            Log.e("daehwan", it.toString())
//                mainListAdapter.submitData(lifecycle,it)
//        }


        binding.rvList.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            adapter = mainListAdapter
        }

        lifecycleScope.launch {
            viewModel.listData.collect {
                mainListAdapter.submitData(it)
            }
        }
    }
}